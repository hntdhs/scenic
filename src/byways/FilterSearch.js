
import React, { useState } from "react";
import BywayApi from "../api/api";
import SearchForm from "../common/SearchForm";

// search via filters, go to browse by state, go to display and search users


function FilterSearch() {
    
    const [byways, setByways] = useState(null);

    // useEffect for getting the byways? need methods for getting random ones or alphabetical order? or let byways, then for loop with those byways to shuffle, then display 10?
    useEffect(() => {
        async function getByways() {
            let byways = await BywayApi.getAllByways();
            // narrow it here?
            setByways(byways);
        }
        getByways();
    }, []);

    async function search(name) {
        let byways = await BywayApi.getByways(name);
        // is getByways right?
        setByways(byways);
    }

    return (
        <div>
            <SearchForm searchFor={search} />
            {/* byway card mapping - could I do a for loop to display 10 byways here too? */}
        </div>
    );
}

export default FilterSearch;


