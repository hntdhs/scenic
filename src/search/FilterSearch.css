.float-container {
    float: left;
    width: 100%;
}