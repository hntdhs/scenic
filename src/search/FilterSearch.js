
import React, { useState, useEffect } from "react";
import { useHistory } from "react-router-dom";
import BywayApi from "../api/api";
import SearchForm from "./SearchForm";
import BywayCard from "../byways/BywayCard";
import Pagination from "../common/Pagination";
import { useToasts } from 'react-toast-notifications';

// search via filters, go to browse by state, go to display and search users

function FilterSearch() {
    const { addToast } = useToasts();
    const [byways, setByways] = useState([]);
    const [errors, setErrors] = useState([]);
    const [currentByWays, setCurrentByWays] = useState([]);
    const [currentPage, setCurrentPage] = useState(1);
    const [bywaysPerPage] = useState(10);
    
    const [nPages, setNPages] = useState(0); 

    const history = useHistory();
    
    useEffect(() => {
        setNPages(Math.ceil(byways.length / bywaysPerPage))
        const indexOfLastByway = currentPage * bywaysPerPage;
        const indexOfFirstByway = indexOfLastByway - bywaysPerPage;
        const currentByways = byways.slice(indexOfFirstByway, indexOfLastByway)
        setCurrentByWays(currentByways)

    }, [byways, currentPage])
    // every time currentPage changes, the useEffect runs

    useEffect(() => {
        async function getBywaysForDisplay() {
            let byways = await BywayApi.getAllByways();
            setByways(byways);
        }
        getBywaysForDisplay();
    }, []);
    // display error to user - try/catch?

    async function filterByways(filters) {
        try {
            debugger
            let byways = await BywayApi.search(filters);
            setByways(byways);
            if (!byways.length) {
                history.push("/noresults");
            }
        } catch (errors) {
            if (errors.length > 0) {
                addToast(errors[0], { appearance: 'error' });
                // can add error messaging/alerts with Toast before anything is sent to the back, like so:
                //addToast("your search term must be more than 3 characters", { appearance: 'error' });
            }
            setErrors(errors);
        }
    }

    async function search(name) {
        let byways = await BywayApi.search({name});
        setByways(byways);
        if (!byways.length) {
            history.push("/noresults");
            // Alert type="success" messages={["Updated successfully."]} 
        }
    }

    return (
        <div>
            <div>
                <SearchForm searchFor={search} filterByways={filterByways} errors={errors} />
                {/* search and filterByways are defined here, passing them over to SearchForm to make them useable there; don't actually know if errors is necessary to pass, didn't end up using it */}
            </div>
            <div>
                {currentByWays.length > 0
                ? (
                    <div>
                        <div>
                            {currentByWays.map(s => (
                                <BywayCard
                                    key={s.name + s.state}
                                    name={s.name}
                                    image={s.image}
                                    designation={s.designation}
                                />
                            ))}
                        </div>
                        <div>
                            <Pagination
                                nPages={nPages}
                                currentPage={currentPage}
                                setCurrentPage={setCurrentPage}
                            />
                        </div>
                        
                    </div>
                ) : (
                    <h4>no byways in system.</h4>
                )}
            </div>
        </div>
        
    );
}

export default FilterSearch;


