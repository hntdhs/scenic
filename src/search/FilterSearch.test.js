import React from "react";
import { render } from "@testing-library/react";
import FilterSearch from "./FilterSearch.js";
import { ToastProvider } from 'react-toast-notifications';

it("matches snapshot", function () {
    const { asFragment } = render(<ToastProvider><FilterSearch /></ToastProvider>);
    expect(asFragment()).toMatchSnapshot();
  });