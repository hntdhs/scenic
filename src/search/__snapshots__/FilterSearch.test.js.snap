// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div
        class="float-container"
      >
        <div
          class="container text-center"
          id="app-name"
        >
          <h1>
            NATIONAL SCENIC BYWAYS
          </h1>
        </div>
        <div
          class="float-child"
        >
          <div
            class="byway-name-search"
          >
            <form>
              <p>
                Search byways by name
              </p>
              <label
                for="searchTerm"
              >
                Byway Name:
              </label>
              <input
                name="searchTerm"
                placeholder="Enter search term.."
                value=""
              />
              <div>
                <button
                  class="btn btn-success btn-lg"
                  type="submit"
                >
                  Submit
                </button>
              </div>
            </form>
          </div>
          <div
            class="not-name-searches"
          >
            <form>
              <div
                class="length-search"
              >
                <p>
                  Search byways by length. The longest of America's BywaysÂ® is the Alaska Marine Highway at 3500 miles!
                </p>
                <div
                  class="input-group input-group-lg"
                >
                  <label
                    class="input-group-text"
                    for="minLength"
                    id="inputGroup-sizing-lg"
                  >
                    Minimum Byway Length:
                  </label>
                  <input
                    name="minLength"
                    placeholder="how many miles?"
                    value="0"
                  />
                </div>
                <div
                  class="input-group input-group-lg"
                >
                  <label
                    class="input-group-text"
                    for="maxLength"
                    id="inputGroup-sizing-lg"
                  >
                    Maximum Byway Length:
                  </label>
                  <input
                    name="maxLength"
                    placeholder="how many miles?"
                    value=""
                  />
                </div>
              </div>
              <div
                class="geo-features-search"
              >
                <p>
                  Search byways by geographic features. You'll see byways that contain all entered criteria.
                </p>
                <div>
                  <div
                    class="form-check form-check-inline"
                  >
                    <label
                      class="form-check-label"
                      for="canyon"
                    >
                      canyon
                    </label>
                    <div>
                      <label>
                        <input
                          class="form-check-input"
                          name="canyon"
                          type="checkbox"
                          value="canyon"
                        />
                      </label>
                    </div>
                  </div>
                  <div
                    class="form-check form-check-inline"
                  >
                    <label
                      class="form-check-label"
                      for="coastline"
                    >
                      coastline
                    </label>
                    <div>
                      <label>
                        <input
                          class="form-check-input"
                          name="coastline"
                          type="checkbox"
                          value="coastline"
                        />
                      </label>
                    </div>
                  </div>
                  <div
                    class="form-check form-check-inline"
                  >
                    <label
                      class="form-check-label"
                      for="desert"
                    >
                      desert
                    </label>
                    <div>
                      <label>
                        <input
                          class="form-check-input"
                          name="desert"
                          type="checkbox"
                          value="desert"
                        />
                      </label>
                    </div>
                  </div>
                  <div
                    class="form-check form-check-inline"
                  >
                    <label
                      class="form-check-label"
                      for="forest"
                    >
                      forest
                    </label>
                    <div>
                      <label>
                        <input
                          class="form-check-input"
                          name="forest"
                          type="checkbox"
                          value="forest"
                        />
                      </label>
                    </div>
                  </div>
                  <div
                    class="form-check form-check-inline"
                  >
                    <label
                      class="form-check-label"
                      for="grasslands"
                    >
                      grasslands
                    </label>
                    <div>
                      <label>
                        <input
                          class="form-check-input"
                          name="grasslands"
                          type="checkbox"
                          value="grasslands"
                        />
                      </label>
                    </div>
                  </div>
                  <div
                    class="form-check form-check-inline"
                  >
                    <label
                      class="form-check-label"
                      for="lake"
                    >
                      lake
                    </label>
                    <div>
                      <label>
                        <input
                          class="form-check-input"
                          name="lake"
                          type="checkbox"
                          value="lake"
                        />
                      </label>
                    </div>
                  </div>
                  <div
                    class="form-check form-check-inline"
                  >
                    <label
                      class="form-check-label"
                      for="mountains"
                    >
                      mountains
                    </label>
                    <div>
                      <label>
                        <input
                          class="form-check-input"
                          name="mountains"
                          type="checkbox"
                          value="mountains"
                        />
                      </label>
                    </div>
                  </div>
                  <div
                    class="form-check form-check-inline"
                  >
                    <label
                      class="form-check-label"
                      for="swamp"
                    >
                      swamp
                    </label>
                    <div>
                      <label>
                        <input
                          class="form-check-input"
                          name="swamp"
                          type="checkbox"
                          value="swamp"
                        />
                      </label>
                    </div>
                  </div>
                  <div
                    class="form-check form-check-inline"
                  >
                    <label
                      class="form-check-label"
                      for="river"
                    >
                      river
                    </label>
                    <div>
                      <label>
                        <input
                          class="form-check-input"
                          name="river"
                          type="checkbox"
                          value="river"
                        />
                      </label>
                    </div>
                  </div>
                  <div
                    class="form-check form-check-inline"
                  >
                    <label
                      class="form-check-label"
                      for="urban"
                    >
                      urban
                    </label>
                    <div>
                      <label>
                        <input
                          class="form-check-input"
                          name="urban"
                          type="checkbox"
                          value="urban"
                        />
                      </label>
                    </div>
                  </div>
                </div>
                <p>
                  Search by official designation
                </p>
                <div>
                  <div
                    class="link"
                  >
                    <a
                      href="/designations-explanation"
                    >
                      What are these?
                    </a>
                  </div>
                  <div
                    class="form-check form-check-inline"
                  >
                    <label
                      for="All-American Road"
                    >
                      All-American Road
                    </label>
                    <div>
                      <label>
                        <input
                          class="form-check-input"
                          name="All-American Road"
                          type="checkbox"
                          value="All-American Road"
                        />
                      </label>
                    </div>
                  </div>
                  <div
                    class="form-check form-check-inline"
                  >
                    <label
                      for="National Scenic Byway"
                    >
                      National Scenic Byway
                    </label>
                    <div>
                      <label>
                        <input
                          class="form-check-input"
                          name="National Scenic Byway"
                          type="checkbox"
                          value="National Scenic Byway"
                        />
                      </label>
                    </div>
                  </div>
                </div>
                <div>
                  <button
                    class="btn btn-success btn-lg"
                    type="submit"
                  >
                    Submit
                  </button>
                </div>
              </div>
            </form>
          </div>
        </div>
        <div
          class="float-child"
        >
          <img
            alt="road through forest"
            class="forest"
            src="https://images.unsplash.com/photo-1592743263126-bb241ee76ac7?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxzZWFyY2h8Mnx8YmVhdXRpZnVsJTIwc2NlbmVyeXxlbnwwfHwwfHx8MA%3D%3D&w=1000&q=80"
          />
        </div>
      </div>
    </div>
    <div>
      <h4>
        no byways in system.
      </h4>
    </div>
  </div>
</DocumentFragment>
`;
